src/middlewared/middlewared/plugins/crypto.py

```diff
@@ -19,80 +24,226 @@
 RE_CERTIFICATE = re.compile(r"(-{5}BEGIN[\s\w]+-{5}[^-]+-{5}END[\s\w]+-{5})+", re.M | re.S)
 
 
-class CertificateService(CRUDService):
+def get_cert_info_from_data(data):
+    cert_info_keys = ['key_length', 'country', 'state', 'city', 'organization', 'common',
+                      'san', 'serial', 'email', 'lifetime', 'digest_algorithm']
+    return {key: data.get(key) for key in cert_info_keys}
 
-    @filterable
-    async def query(self, filters=None, options=None):
-        if options is None:
-            options = {}
-        options['extend'] = 'certificate.cert_extend'
-        return await self.middleware.call('datastore.query', 'system.certificate', filters, options)
 
-    @accepts(
-        Int('certificate_id', required=True)
+async def validate_cert_name(middleware, cert_name, datastore, verrors, name):
+    certs = await middleware.call(
+        'datastore.query',
+        datastore,
+        [('cert_name', '=', cert_name)]
     )
-    async def get_fingerprint(self, certificate_id):
-        certificate_list = await self.query(filters=[('id', '=', certificate_id)])
-        if len(certificate_list) == 0:
-            return None
-        else:
-            cert_certificate = certificate_list[0]['cert_certificate']
+    if certs:
+        verrors.add(
+            name,
+            'A certificate with this name already exists'
+        )
 
-        # getting fingerprint of certificate
-        try:
-            certificate = crypto.load_certificate(
-                crypto.FILETYPE_PEM,
-                cert_certificate
+    if cert_name in ("external", "self-signed", "external - signature pending"):
+        verrors.add(
+            name,
+            '{0} is a reserved internal keyword for Certificate Management'.format(cert_name)
```

Comment:
More of a nitpick but can we use f-string? We are trying to unify across all middleware